src/lib/transition_router/transition_router.ml

```diff
@@ -102,15 +74,231 @@ let start_bootstrap_controller ~logger ~trust_system ~verifier ~network
+        is_transition_for_bootstrap ~logger frontier
+          (best_tip |> Envelope.Incoming.data)
+      then
+        let initial_root_transition =
+          Transition_frontier.(Breadcrumb.validated_transition (root frontier))
+        in
+        let%map () = Transition_frontier.close frontier in
+        start_bootstrap_controller ~logger ~trust_system ~verifier ~network
+          ~time_controller ~proposer_transition_reader
+          ~verified_transition_writer ~clear_reader ~transition_reader_ref
+          ~consensus_local_state ~transition_writer_ref ~frontier_w
+          ~persistent_root ~persistent_frontier ~initial_root_transition
+      else
+        let root = Transition_frontier.root frontier in
+        let%map () =
+          match
+            Consensus.Hooks.required_local_state_sync
+              ~consensus_state:
+                (Transition_frontier.Breadcrumb.consensus_state root)
+              ~local_state:consensus_local_state
+          with
+          | None ->
+              Deferred.unit
+          | Some sync_jobs -> (
+              match%map
+                Consensus.Hooks.sync_local_state
+                  ~local_state:consensus_local_state ~logger ~trust_system
+                  ~random_peers:(Coda_networking.random_peers network)
+                  ~query_peer:
+                    { Consensus.Hooks.Rpcs.query=
+                        (fun peer rpc query ->
+                          Coda_networking.(
+                            query_peer network peer (Rpcs.Consensus_rpc rpc)
+                              query) ) }
+                  sync_jobs
+              with
+              | Error e ->
+                  failwith ("Local state sync failed: " ^ Error.to_string_hum e)
+              | Ok () ->
+                  () )
+        in
+        start_transition_frontier_controller ~logger ~trust_system ~verifier
+          ~network ~time_controller ~proposer_transition_reader
+          ~verified_transition_writer ~clear_reader
+          ~collected_transitions:[best_tip] ~transition_reader_ref
+          ~transition_writer_ref ~frontier_w frontier
+
+let wait_till_genesis ~logger ~time_controller =
+  let module Time = Coda_base.Block_time in
+  let now = Time.now time_controller in
+  try Consensus.Hooks.is_genesis now |> ignore
+  with Invalid_argument _ ->
+    let time_till_genesis =
+      Time.diff Consensus.Constants.genesis_state_timestamp now
+    in
+    Logger.warn logger ~module_:__MODULE__ ~location:__LOC__
+      ~metadata:
+        [ ( "time_till_genesis"
+          , `Int (Int64.to_int_exn (Time.Span.to_ms time_till_genesis)) ) ]
+      "Node started before genesis: waiting $time_till_genesis milliseconds \
+       before running transition router" ;
+    let seconds_to_wait = 30 in
+    let milliseconds_to_wait = Int64.of_int_exn (seconds_to_wait * 1000) in
+    let rec wait_loop tm =
+      if Int64.(tm <= zero) then ()
+      else (
+        Core.Unix.sleep seconds_to_wait ;
```

Comment:
I think you should use async here instead of a sleep loop. Sleeping like this will cause delays when making any requests to the daemon via RPC or GraphQL, which should be accessible even when waiting till genesis. Under this implementation, you may not start exactly at genesis either since you are only checking in 30 second intervals. I would recommend using the `Block_time.Timeout` module (https://github.com/CodaProtocol/coda/blob/develop/src/lib/coda_base/timeout.ml#L52) and create a timeout on an infinite log loop. Then you can just bind on the timeout and it will be resolved precisely at genesis.